### PR TITLE
qrencode

### PR DESCRIPTION
**Pull request by jperon**
This PR is necessary for https://github.com/koreader/koreader/pull/6844.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader-base/1227)

### CODE REVIEW

File: ffi/qrencode.lua [hunk 1]
```diff
@@ -0,0 +1,1325 @@
...
+--- Each step is of course more or less complex and needs further description
+
+--- Helper functions
+--- ================
+---
+--- We start with some helper functions
+
+-- To calculate xor we need to do that bitwise. This helper table speeds up the num-to-bit
+-- part a bit (no pun intended)
+local cclxvi = {[0] = {0,0,0,0,0,0,0,0}, {1,0,0,0,0,0,0,0}, {0,1,0,0,0,0,0,0}, {1,1,0,0,0,0,0,0},
```
Comment (pazos) [inline ffi/qrencode.lua:50]: this table isn't needed with luajit bitop

File: ffi/qrencode.lua [hunk 2]
```diff
@@ -0,0 +1,1325 @@
...
+{0,0,1,0,0,1,1,1}, {1,0,1,0,0,1,1,1}, {0,1,1,0,0,1,1,1}, {1,1,1,0,0,1,1,1},
+{0,0,0,1,0,1,1,1}, {1,0,0,1,0,1,1,1}, {0,1,0,1,0,1,1,1}, {1,1,0,1,0,1,1,1},
+{0,0,1,1,0,1,1,1}, {1,0,1,1,0,1,1,1}, {0,1,1,1,0,1,1,1}, {1,1,1,1,0,1,1,1},
+{0,0,0,0,1,1,1,1}, {1,0,0,0,1,1,1,1}, {0,1,0,0,1,1,1,1}, {1,1,0,0,1,1,1,1},
+{0,0,1,0,1,1,1,1}, {1,0,1,0,1,1,1,1}, {0,1,1,0,1,1,1,1}, {1,1,1,0,1,1,1,1},
+{0,0,0,1,1,1,1,1}, {1,0,0,1,1,1,1,1}, {0,1,0,1,1,1,1,1}, {1,1,0,1,1,1,1,1},
+{0,0,1,1,1,1,1,1}, {1,0,1,1,1,1,1,1}, {0,1,1,1,1,1,1,1}, {1,1,1,1,1,1,1,1}}
+
+-- Return a number that is the result of interpreting the table tbl (msb first)
+local function tbl_to_number(tbl)
```
Comment (pazos) [inline ffi/qrencode.lua:116]: same?

File: ffi/qrencode.lua [hunk 3]
```diff
@@ -0,0 +1,1325 @@
...
+    local power = 1
+    for i = 1, n do
+        rslt = rslt + tbl[i]*power
+        power = power*2
+    end
+    return rslt
+end
+
+-- Calculate bitwise xor of bytes m and n. 0 <= m,n <= 256.
+local function bit_xor(m, n)
```
Comment (pazos) [inline ffi/qrencode.lua:128]: probably will be faster (and less boilerplate) using bit.xor from luajit bitop

File: ffi/qrencode.lua [hunk 4]
```diff
@@ -0,0 +1,1325 @@
...
+            tbl[i] = 1
+        else
+            tbl[i] = 0
+        end
+    end
+    return tbl_to_number(tbl)
+end
+
+-- Return the binary representation of the number x with the width of `digits`.
+local function binary(x,digits)
```
Comment (pazos) [inline ffi/qrencode.lua:143]: probably can be simplified using bitop

**Review by pazos — COMMENTED** [review]
At a first glance it looks that some functions can be simplified or avoided by using luajit bitop module.

I would like to try before merging this as is.

Could you write an unit test that does some encode / decode and asserts that both match?. That would make my tinkering a lot easier :)

If you don't want to mess with unit test I can submit a PR to your branch.

**Comment by jperon** [comment]
@pazos I've tried to write a unit test by looking at other ones, but I'm not sure it's good at all…

As `qrencode` doesn't have decode functions, I only ensured that 'test' gets encoded the same.

**Comment by pazos** [comment]
thanks @jperon. I added your test in a for loop.

The binary helper function can be written as

```lua
local function bin(n, digits)
    local t = {}
    for i = 1, 32 do
        n = bit.rol(n, 1)
        table.insert(t, bit.band(n, 1))
    end
    local s = string.gsub(table.concat(t), "^0*(.*)$","%1")
    local fmtstring = string.format("%%%ds", digits)
    local ret = string.format(fmtstring,s)
    return string.gsub(ret, " ", "0")
end
```

which is a bit easier to read than the binary function but not faster.

The bit_xor helper function can be replaced by bit.bxor. It is a bit faster but not as faster as I could expect

in a 10000 loop results are `(12314.63 ms total)` using original function and `(11286.24 ms total)` using bit.xor.

So, being marginally faster the only benefit from including it is less boilerplate code, as in the diff

```diff
diff --git a/ffi/qrencode.lua b/ffi/qrencode.lua
index 9d96dee..86f86d0 100644
--- a/ffi/qrencode.lua
+++ b/ffi/qrencode.lua
@@ -45,99 +45,9 @@
 ---
 --- We start with some helper functions
 
--- To calculate xor we need to do that bitwise. This helper table speeds up the num-to-bit
--- part a bit (no pun intended)
-local cclxvi = {[0] = {0,0,0,0,0,0,0,0}, {1,0,0,0,0,0,0,0}, {0,1,0,0,0,0,0,0}, {1,1,0,0,0,0,0,0},
-{0,0,1,0,0,0,0,0}, {1,0,1,0,0,0,0,0}, {0,1,1,0,0,0,0,0}, {1,1,1,0,0,0,0,0},
-{0,0,0,1,0,0,0,0}, {1,0,0,1,0,0,0,0}, {0,1,0,1,0,0,0,0}, {1,1,0,1,0,0,0,0},
-{0,0,1,1,0,0,0,0}, {1,0,1,1,0,0,0,0}, {0,1,1,1,0,0,0,0}, {1,1,1,1,0,0,0,0},
-{0,0,0,0,1,0,0,0}, {1,0,0,0,1,0,0,0}, {0,1,0,0,1,0,0,0}, {1,1,0,0,1,0,0,0},
-{0,0,1,0,1,0,0,0}, {1,0,1,0,1,0,0,0}, {0,1,1,0,1,0,0,0}, {1,1,1,0,1,0,0,0},
-{0,0,0,1,1,0,0,0}, {1,0,0,1,1,0,0,0}, {0,1,0,1,1,0,0,0}, {1,1,0,1,1,0,0,0},
-{0,0,1,1,1,0,0,0}, {1,0,1,1,1,0,0,0}, {0,1,1,1,1,0,0,0}, {1,1,1,1,1,0,0,0},
-{0,0,0,0,0,1,0,0}, {1,0,0,0,0,1,0,0}, {0,1,0,0,0,1,0,0}, {1,1,0,0,0,1,0,0},
-{0,0,1,0,0,1,0,0}, {1,0,1,0,0,1,0,0}, {0,1,1,0,0,1,0,0}, {1,1,1,0,0,1,0,0},
-{0,0,0,1,0,1,0,0}, {1,0,0,1,0,1,0,0}, {0,1,0,1,0,1,0,0}, {1,1,0,1,0,1,0,0},
-{0,0,1,1,0,1,0,0}, {1,0,1,1,0,1,0,0}, {0,1,1,1,0,1,0,0}, {1,1,1,1,0,1,0,0},
-{0,0,0,0,1,1,0,0}, {1,0,0,0,1,1,0,0}, {0,1,0,0,1,1,0,0}, {1,1,0,0,1,1,0,0},
-{0,0,1,0,1,1,0,0}, {1,0,1,0,1,1,0,0}, {0,1,1,0,1,1,0,0}, {1,1,1,0,1,1,0,0},
-{0,0,0,1,1,1,0,0}, {1,0,0,1,1,1,0,0}, {0,1,0,1,1,1,0,0}, {1,1,0,1,1,1,0,0},
-{0,0,1,1,1,1,0,0}, {1,0,1,1,1,1,0,0}, {0,1,1,1,1,1,0,0}, {1,1,1,1,1,1,0,0},
-{0,0,0,0,0,0,1,0}, {1,0,0,0,0,0,1,0}, {0,1,0,0,0,0,1,0}, {1,1,0,0,0,0,1,0},
-{0,0,1,0,0,0,1,0}, {1,0,1,0,0,0,1,0}, {0,1,1,0,0,0,1,0}, {1,1,1,0,0,0,1,0},
-{0,0,0,1,0,0,1,0}, {1,0,0,1,0,0,1,0}, {0,1,0,1,0,0,1,0}, {1,1,0,1,0,0,1,0},
-{0,0,1,1,0,0,1,0}, {1,0,1,1,0,0,1,0}, {0,1,1,1,0,0,1,0}, {1,1,1,1,0,0,1,0},
-{0,0,0,0,1,0,1,0}, {1,0,0,0,1,0,1,0}, {0,1,0,0,1,0,1,0}, {1,1,0,0,1,0,1,0},
-{0,0,1,0,1,0,1,0}, {1,0,1,0,1,0,1,0}, {0,1,1,0,1,0,1,0}, {1,1,1,0,1,0,1,0},
-{0,0,0,1,1,0,1,0}, {1,0,0,1,1,0,1,0}, {0,1,0,1,1,0,1,0}, {1,1,0,1,1,0,1,0},
-{0,0,1,1,1,0,1,0}, {1,0,1,1,1,0,1,0}, {0,1,1,1,1,0,1,0}, {1,1,1,1,1,0,1,0},
-{0,0,0,0,0,1,1,0}, {1,0,0,0,0,1,1,0}, {0,1,0,0,0,1,1,0}, {1,1,0,0,0,1,1,0},
-{0,0,1,0,0,1,1,0}, {1,0,1,0,0,1,1,0}, {0,1,1,0,0,1,1,0}, {1,1,1,0,0,1,1,0},
-{0,0,0,1,0,1,1,0}, {1,0,0,1,0,1,1,0}, {0,1,0,1,0,1,1,0}, {1,1,0,1,0,1,1,0},
-{0,0,1,1,0,1,1,0}, {1,0,1,1,0,1,1,0}, {0,1,1,1,0,1,1,0}, {1,1,1,1,0,1,1,0},
-{0,0,0,0,1,1,1,0}, {1,0,0,0,1,1,1,0}, {0,1,0,0,1,1,1,0}, {1,1,0,0,1,1,1,0},
-{0,0,1,0,1,1,1,0}, {1,0,1,0,1,1,1,0}, {0,1,1,0,1,1,1,0}, {1,1,1,0,1,1,1,0},
-{0,0,0,1,1,1,1,0}, {1,0,0,1,1,1,1,0}, {0,1,0,1,1,1,1,0}, {1,1,0,1,1,1,1,0},
-{0,0,1,1,1,1,1,0}, {1,0,1,1,1,1,1,0}, {0,1,1,1,1,1,1,0}, {1,1,1,1,1,1,1,0},
-{0,0,0,0,0,0,0,1}, {1,0,0,0,0,0,0,1}, {0,1,0,0,0,0,0,1}, {1,1,0,0,0,0,0,1},
-{0,0,1,0,0,0,0,1}, {1,0,1,0,0,0,0,1}, {0,1,1,0,0,0,0,1}, {1,1,1,0,0,0,0,1},
-{0,0,0,1,0,0,0,1}, {1,0,0,1,0,0,0,1}, {0,1,0,1,0,0,0,1}, {1,1,0,1,0,0,0,1},
-{0,0,1,1,0,0,0,1}, {1,0,1,1,0,0,0,1}, {0,1,1,1,0,0,0,1}, {1,1,1,1,0,0,0,1},
-{0,0,0,0,1,0,0,1}, {1,0,0,0,1,0,0,1}, {0,1,0,0,1,0,0,1}, {1,1,0,0,1,0,0,1},
-{0,0,1,0,1,0,0,1}, {1,0,1,0,1,0,0,1}, {0,1,1,0,1,0,0,1}, {1,1,1,0,1,0,0,1},
-{0,0,0,1,1,0,0,1}, {1,0,0,1,1,0,0,1}, {0,1,0,1,1,0,0,1}, {1,1,0,1,1,0,0,1},
-{0,0,1,1,1,0,0,1}, {1,0,1,1,1,0,0,1}, {0,1,1,1,1,0,0,1}, {1,1,1,1,1,0,0,1},
-{0,0,0,0,0,1,0,1}, {1,0,0,0,0,1,0,1}, {0,1,0,0,0,1,0,1}, {1,1,0,0,0,1,0,1},
-{0,0,1,0,0,1,0,1}, {1,0,1,0,0,1,0,1}, {0,1,1,0,0,1,0,1}, {1,1,1,0,0,1,0,1},
-{0,0,0,1,0,1,0,1}, {1,0,0,1,0,1,0,1}, {0,1,0,1,0,1,0,1}, {1,1,0,1,0,1,0,1},
-{0,0,1,1,0,1,0,1}, {1,0,1,1,0,1,0,1}, {0,1,1,1,0,1,0,1}, {1,1,1,1,0,1,0,1},
-{0,0,0,0,1,1,0,1}, {1,0,0,0,1,1,0,1}, {0,1,0,0,1,1,0,1}, {1,1,0,0,1,1,0,1},
-{0,0,1,0,1,1,0,1}, {1,0,1,0,1,1,0,1}, {0,1,1,0,1,1,0,1}, {1,1,1,0,1,1,0,1},
-{0,0,0,1,1,1,0,1}, {1,0,0,1,1,1,0,1}, {0,1,0,1,1,1,0,1}, {1,1,0,1,1,1,0,1},
-{0,0,1,1,1,1,0,1}, {1,0,1,1,1,1,0,1}, {0,1,1,1,1,1,0,1}, {1,1,1,1,1,1,0,1},
-{0,0,0,0,0,0,1,1}, {1,0,0,0,0,0,1,1}, {0,1,0,0,0,0,1,1}, {1,1,0,0,0,0,1,1},
-{0,0,1,0,0,0,1,1}, {1,0,1,0,0,0,1,1}, {0,1,1,0,0,0,1,1}, {1,1,1,0,0,0,1,1},
-{0,0,0,1,0,0,1,1}, {1,0,0,1,0,0,1,1}, {0,1,0,1,0,0,1,1}, {1,1,0,1,0,0,1,1},
-{0,0,1,1,0,0,1,1}, {1,0,1,1,0,0,1,1}, {0,1,1,1,0,0,1,1}, {1,1,1,1,0,0,1,1},
-{0,0,0,0,1,0,1,1}, {1,0,0,0,1,0,1,1}, {0,1,0,0,1,0,1,1}, {1,1,0,0,1,0,1,1},
-{0,0,1,0,1,0,1,1}, {1,0,1,0,1,0,1,1}, {0,1,1,0,1,0,1,1}, {1,1,1,0,1,0,1,1},
-{0,0,0,1,1,0,1,1}, {1,0,0,1,1,0,1,1}, {0,1,0,1,1,0,1,1}, {1,1,0,1,1,0,1,1},
-{0,0,1,1,1,0,1,1}, {1,0,1,1,1,0,1,1}, {0,1,1,1,1,0,1,1}, {1,1,1,1,1,0,1,1},
-{0,0,0,0,0,1,1,1}, {1,0,0,0,0,1,1,1}, {0,1,0,0,0,1,1,1}, {1,1,0,0,0,1,1,1},
-{0,0,1,0,0,1,1,1}, {1,0,1,0,0,1,1,1}, {0,1,1,0,0,1,1,1}, {1,1,1,0,0,1,1,1},
-{0,0,0,1,0,1,1,1}, {1,0,0,1,0,1,1,1}, {0,1,0,1,0,1,1,1}, {1,1,0,1,0,1,1,1},
-{0,0,1,1,0,1,1,1}, {1,0,1,1,0,1,1,1}, {0,1,1,1,0,1,1,1}, {1,1,1,1,0,1,1,1},
-{0,0,0,0,1,1,1,1}, {1,0,0,0,1,1,1,1}, {0,1,0,0,1,1,1,1}, {1,1,0,0,1,1,1,1},
-{0,0,1,0,1,1,1,1}, {1,0,1,0,1,1,1,1}, {0,1,1,0,1,1,1,1}, {1,1,1,0,1,1,1,1},
-{0,0,0,1,1,1,1,1}, {1,0,0,1,1,1,1,1}, {0,1,0,1,1,1,1,1}, {1,1,0,1,1,1,1,1},
-{0,0,1,1,1,1,1,1}, {1,0,1,1,1,1,1,1}, {0,1,1,1,1,1,1,1}, {1,1,1,1,1,1,1,1}}
-
--- Return a number that is the result of interpreting the table tbl (msb first)
-local function tbl_to_number(tbl)
-    local n = #tbl
-    local rslt = 0
-    local power = 1
-    for i = 1, n do
-        rslt = rslt + tbl[i]*power
-        power = power*2
-    end
-    return rslt
-end
-
--- Calculate bitwise xor of bytes m and n. 0 <= m,n <= 256.
-local function bit_xor(m, n)
-    local tbl_m = cclxvi[m]
-    local tbl_n = cclxvi[n]
-    local tbl = {}
-    for i = 1, 8 do
-        if(tbl_m[i] ~= tbl_n[i]) then
-            tbl[i] = 1
-        else
-            tbl[i] = 0
-        end
-    end
-    return tbl_to_number(tbl)
-end
+--- slightly modified to work with LuaJIT bitop
+local bit = require("bit")
+local bxor = bit.bxor
 
 -- Return the binary representation of the number x with the width of `digits`.
 local function binary(x,digits)
@@ -589,7 +499,7 @@ local function calculate_error_correction(data,num_ec_codewords)
 
         tmp = {}
         for i=highest_exponent,0,-1 do
-            tmp[i] = bit_xor(gp_int[i],mp_int[i])
+            tmp[i] = bxor(gp_int[i],mp_int[i])
         end
         -- remove leading 0's
         he = highest_exponent
@@ -1322,4 +1232,4 @@ end
 
 return {
     qrcode = qrcode
-}
\ No newline at end of file
+}
```

I'm fine leaving the file as is (it is easier to follow upstream). :+1:

**Review by pazos — APPROVED** [review]

**Comment by jperon** [comment]
Thanks @pazos. So is there anything else to do from my part ?

**Comment by pazos** [comment]
> Thanks @pazos. So is there anything else to do from my part ?

Nope. LGTM. Also the unit test will be useful in the future if someone attempts to port https://fukuchi.org/works/qrencode while keeping the same output.

On https://github.com/koreader/koreader/pull/6844 you'll need to remove the file, checkout base on master branch, add base and commit the bump.

**Pull request merged**